core/trino-main/src/main/java/io/trino/operator/TableFunctionOperator.java

```diff
@@ -0,0 +1,566 @@
+
+            if (page.getPositionCount() - groupEnd > 0) {
+                // the remaining prt of the page contains the next group
+                return groupEnd;
+            }
+            // page fully consumed: it contains the current group only
+            return page.getPositionCount();
+        }
+
+        // we had previous results buffered, but the remaining page starts with new group values
+        return startPosition;
+    }
+
+    private static void sortCurrentGroup(PagesIndex pagesIndex, HashStrategies hashStrategies)
+    {
+        PagesHashStrategy preSortedStrategy = hashStrategies.preSortedStrategy;
+        List<Integer> remainingPartitionAndSortChannels = hashStrategies.remainingPartitionAndSortChannels;
+        List<SortOrder> remainingSortOrders = hashStrategies.remainingSortOrders;
+
+        if (pagesIndex.getPositionCount() > 1 && !remainingPartitionAndSortChannels.isEmpty()) {
+            int startPosition = 0;
+            while (startPosition < pagesIndex.getPositionCount()) {
+                int endPosition = findGroupEnd(pagesIndex, preSortedStrategy, startPosition);
+                pagesIndex.sort(remainingPartitionAndSortChannels, remainingSortOrders, startPosition, endPosition);
+                startPosition = endPosition;
+            }
+        }
+    }
+
+    // Assumes input grouped on relevant pagesHashStrategy columns
+    private static int findGroupEnd(Page page, PagesHashStrategy pagesHashStrategy, int startPosition)
+    {
+        checkArgument(page.getPositionCount() > 0, "Must have at least one position");
+        checkPositionIndex(startPosition, page.getPositionCount(), "startPosition out of bounds");
+
+        return findEndPosition(startPosition, page.getPositionCount(), (firstPosition, secondPosition) -> pagesHashStrategy.rowNotDistinctFromRow(firstPosition, page, secondPosition, page));
+    }
+
+    // Assumes input grouped on relevant pagesHashStrategy columns
+    private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHashStrategy, int startPosition)
+    {
+        checkArgument(pagesIndex.getPositionCount() > 0, "Must have at least one position");
+        checkPositionIndex(startPosition, pagesIndex.getPositionCount(), "startPosition out of bounds");
+
+        return findEndPosition(startPosition, pagesIndex.getPositionCount(), (firstPosition, secondPosition) -> pagesIndex.positionNotDistinctFromPosition(pagesHashStrategy, firstPosition, secondPosition));
+    }
+
+    /**
+     * @param startPosition - inclusive
+     * @param endPosition - exclusive
+     * @param comparator - returns true if positions given as parameters are equal
+     * @return the end of the group position exclusive (the position the very next group starts)
+     */
+    @VisibleForTesting
+    static int findEndPosition(int startPosition, int endPosition, BiPredicate<Integer, Integer> comparator)
```

Comment:
`BiPredicate<Integer, Integer>` will cause boxing of the values. It might be worth creating a custom interface for the comparator that takes two primitive ints.